### ISSUE TITLE
Simplify project/profile membership

### ISSUE THREAD

**Issue by jace** [hasgeek/funnel#371]
Users can be associated with profiles and projects via one of several models:

1. `Team` (to Profile or Project), wherein membership is synced from Lastuser, but unreliably.
2. The erstwhile `UserGroup`, removed in #344 and #370.
3. `Rsvp`, recording a yes/no/maybe status
4. `Participant`, in the convoluted dual chain of Project → Event → Attendee → Participant and also Project → Event → Participant → Attendee.

This ticket proposes that both the `Team` and `Rsvp` models be dumped in favour of a new `Member` model (which could also be a new name for `Rsvp`). Members are attached to Projects, can have roles (admin, editor, reviewer, none), and can have a RSVP status, although that should ideally move down to be attached to the Event model.

Roles in projects will be read entirely from the Member model. The Member model will initially be populated from existing project teams. Profiles will also lose the admin team feature, restricting admin access to Owners from Lastuser.

**Comment by jace** [comment]
`Participant` and associated models should be discussed in a separate ticket.

**Comment by jace** [comment]
I've run into a hitch with the `Member` model. It makes sense to think of it as a join between `Profile` and User`, defined with a composite primary key.

However, if we are to add members to the profile and give them roles in the project, then we'll need a `ProjectMember` that joins `Project` with `Member`, which we can't because `Member` doesn't have a primary key. The join will actually be between Project and Profile+User which breaks normalisation by having two links to Profile. We will have to make profile membership and project membership distinct concepts.

**Comment by jace** [comment]
Since projects can be moved between profiles, independent membership makes sense. It just means we will need to duplicate UI in two places, potentially confusing users.

**Comment by jace** [comment]
Membership also has a timeline that needs to be recorded: when did the person become a member? When did they get specific roles? When was their membership revoked? By whom? We need to record membership event as well.

These can't be recorded against the `Member` record if it doesn't have a singular primary key. On the other hand, there's no particular need to record against Member either -- events could be recorded against user vs project/profile similar to members themself. The events will survive deleted memberships.

**Comment by jace** [comment]
Kickstarter uses a Pledge model. If a backer has an existing pledge against a project, they can't make a new pledge. But if a backer cancels a pledge and then pledges again, the cancellation will remain on the records.

This offers us an insight: make membership records immutable. Roles cannot be changed once a record is created. However, the membership can be replaced with a new membership with a new set of roles. The old one is marked as deactivated and is not used for enumeration, but is available to examine the timeline of role changes.

**Comment by jace** [comment]
A partial unique index can be used to enforce only one active membership record at a time. https://stackoverflow.com/a/27979223/78903

**Comment by jace** [comment]
If the model tracks a timeline, then it's more accurately called a "membership" record and not "member".

**Comment by jace** [comment]
I'm conflicted on whether the `Rsvp` model should be merged into `ProjectMembership`. Changes in status need to be recorded on a timeline, so it makes logical sense to include it. On the other hand, RSVP is complicated. Does a user's response indicate actual intent? Are four states (yes, no, maybe, unknown) sufficient? Do we need something else for the organizer's guess, because after all accurate status tracking matters for the project's headcount?

Most importantly, how do we iterate on this in the future? The membership model is meant to be immutable, with a new record added everytime the content changes. How does that work when the data model itself changes as new states are added or deleted? Note that we have the same problem when membership roles are added or deleted.

**Comment by jace** [comment]
When a membership is revoked, it can be because it was replaced with a new membership (a technical revocation), or because it was entirely revoked (a product revocation). These are currently distinguished on the basis of whether a new membership record began at the same time. We may need something more substantial such as a state column containing one of:

1. Active
2. Replaced
3. Terminated

**Comment by jace** [comment]
Another approach: have a membership id column with a random non-unique value. The value is replicated when a membership is replaced, but generated afresh otherwise. A validator on the column must ensure old values cannot be recycled unless the value is from the currently active membership record.

**Comment by jace** [comment]
Using a membership id makes it possible to reconstruct a timeline of periods of membership using something like this:

```sql
SELECT membershipid, MIN(granted_at), MAX(revoked_at) FROM project_membership GROUP BY membershipid
```

**Comment by jace** [comment]
Instead of a random value, the membership id could be the UUID of the first record in a chain. This way the first is identifiable from matching UUID and membership id, and the last from its active state.

**Comment by jace** [comment]
Why not split the membership into `(Project|Profile)Membership` and `(Project|Profile)RoleSet` models? Both can be revoked as outlined above. Changing roles won't break the membership record. The roleset can be invisible outside the membership model, with Membership operating as a passthrough: `membership_obj.roles.is_admin` etc.

**Comment by jace** [comment]
We can take this further and drop the membership model entirely. Let each role be its own table using the revocable record model. `ProfileAdminRole`, `ProjectReviewerRole`, etc. A user's past involvement can be discovered on a per-role basis. Since each role has distinct UI behaviour, this can help not forcefit them with each other (for example, RSVP status with reviewer status), but it will also make it harder to discover overall involvement as multiple tables will need to be queried.

**Comment by jace** [comment]
Option:

```python
class ProjectMembership(UuidMixin, NoIdMixin, db.Model):
	__tablename__ = 'project_membership'
    # A `uuid` column is added by `UuidMixin`
    project_id = db.Column(None, db.ForeignKey('project.id'), primary_key=True)
    user_id = db.Column(None, db.ForeignKey('user.id'), primary_key=True, index=True)


class ReviewerRole(UuidMixin, BaseMixin, db.Model):
    __tablename__ = 'reviewer_role'
    member_id = db.Column(None, db.ForeignKey('project_membership.uuid'), nullable=False)
    granted_at = db.Column(db.DateTime, default=db.func.utcnow(), nullable=False)
    revoked_at = db.Column(db.DateTime, nullable=True)
    __table_args__ = (db.Index('review_role_active', 'member_id', unique=True, postgresql_where=revoked_at.isnot(None)),)
```

In this model, members can be easily enumerated from the membership model, but roles are attached to it from role-specific tables.

Problem: a revoked member can only be identified by scanning all role tables. This is expensive.

**Comment by jace** [comment]
This is now the subject of PR #401.

**Comment by jace** [comment]
The design as it stands today:

1. A "membership" is a bundle of related roles assigned to a user on a given target like a Project or a Profile.

2. The roles in a membership are immutable. They cannot be changed once assigned. However, the membership itself can be revoked and a new membership can be issued in its place.

3. There will be multiple membership models, each representing a different bundle of roles and targets. PR #401 currently includes a `ProfileAdminMembership` and `ProjectCrewMembership`. The latter describes `editor`, `concierge` and `usher` roles, but does not include `reviewer`, `participant`, `proposer` or `speaker`. These should be described in other membership models.

4. Each membership model implements its own logic for defining roles. In `ProfileAdminMembership`, `admin` role is granted by default while `owner` role is an optional upgrade. In `ProjectCrewMembership` however, all three roles are distinct and at least one must be granted for the record to be valid. A future `ProjectParticipantMembership` can have an RSVP column for yes/no/maybe/waiting states and treat each of these as a separate role.

5. Because membership has both a start and an end, it's possible to have a timeline of changes. For audit reasons we will want to know whether someone had a particular role at a particular time in the past, and for product development it is useful to know the triggers for a participant changing their RSVP status.

6. **What is not clear yet**: how do the Profile and Project models discover roles for a user? Do they comb through all the known membership models? How do the membership models determine roles on themselves? For instance, someone with `owner` role on Profile can revoke the `admin` role of another user. However, to access `ProfileAdminMembership.revoke()`, the `owner` role has to be known somehow. If the membership model consults the Profile model for roles, it will set off a recursive loop since the Profile model will in turn consult the membership model.

**Comment by jace** [comment]
The membership model can have a property named `offered_roles` which returns a set of roles as granted by the record. This method does not test whether the membership is valid or not. It's simply an abstraction converting from the model's internal representation to the mechanism expected by the roles infrastructure.

1. The `roles_for` methods on Project and Profile can look up the offered roles on active membership records -- although they still need to be aware of the existence of specific membership models. This is iffy because they're defined in separate modules and when a developer adds a new type of membership record they will need to update the corresponding methods in other modules.

2. The `roles_for` methods on membership models can call upstream to the Profile or Project. This will not result in recursive call.

**Comment by jace** [comment]
Is there a possibility of a membership record having more than one user? `offered_roles` will then vary based on the user and should be a method rather than a property. I can't think of an appropriate scenario at the moment however.

**Comment by jace** [comment]
Since `offered_roles` _may_ have a non-trivial lookup cost, it should be a function instead of a property. This leaves the door open for future expansion.

**Comment by jace** [comment]
Invited memberships were not considered in the first draft of PR #401. This review comment addresses how they can be added. https://github.com/hasgeek/funnel/pull/401#pullrequestreview-283391313

**Comment by iambibhas** [comment]
@jace need some clarity into which models will have which roles available through `current_roles`.

`Profile.current_roles.`
- `owner`
- `admin`

`Project.current_roles.`
- `reviewer`
- `concierge`
- `usher`
- `profile_owner` ?
- `profile_admin` ?

`Proposal.current_roles.`
- `author`
- `reviewer`
- `project_<foo>` ?

**Comment by jace** [comment]
This is correct except `editor` instead of `reviewer` in Project. There is an issue to settle: since `editor` is a standard role in Coaster, it is expected to be granted based on other roles. For instance, if the concierge is allowed to edit the project description, then they should get `editor` role.

However, since the only standard roles actually in use are `all`, `anon` and `auth`, we have a choice:

1. Treat `editor` as a specific role along with `concierge` and `usher`.
2. Call it something else and reserve the standard role list.

I'm leaning towards 1.

**Comment by iambibhas** [comment]
1 sounds better to me as well.

**Comment by jace** [comment]
Membership models have been in stable use for a while now, with only proposal membership pending (with a forthcoming PR). Closing this ticket.

**Comment by jace** [comment]
Proposal memberships were released into production use in #1122.
